Agrego guía para emular VMs en macOS con ARM (#74)
sisoputnfrba/docs: docs/guias/herramientas/macos-arm.md

Ahora sí, dale `Save`, y ejecutá la VM dándole al botón de Play. ¡Listo!

## Corregir la configuración de red

Probablemente la configuración de red de la VM no te funcione, porque
QEMU seguramente le asigne a la placa de red un nombre de dispositivo
distinto al que tenía el disco que bajaste.

Para corregirlo, iniciá sesión en la VM, ejecutá `ip addr`, y fijate si
la entrada con nombre del estilo `enp0s1` (quizá ese mismo) tiene una
línea que comience con `inet` y una IP. Si está: fiesta, tenés internet,
ganaste.

Si la interfaz de red no está configurada, editá el archivo
`/etc/netplan/00-installer-config.yml` con tu editor favorito (¡acordate
de usar `sudo`!), y reemplazá en la línea que dice `enp0s3:` el nombre
de dispositivo que te haya listado `ip addr` (_¡dejale el `:` al final,
eh!_).

::: tip

Si no tenés un editor de textos favorito, ejecutá `sudo nano
/etc/netplan/00-installer-config.yml`, poné la contraseña de tu usuario
`utnso`, hacé los cambios al archivo, y después tocá CTRL+X (_Exit_),
después `Y` ("Yes") cuando te pregunta si querés guardar el "buffer", y
por último dale `Enter` para que guarde en la misma ruta que habías

> abierto.

Reviewer: Tenemos una mini guía de nano acá, si te parece oportuno podemos linkearla :D
https://docs.utnso.com.ar/guias/consola/bash#nano

Author: Re sí 👍

tenés una conexión de red funcionando (buscá la línea de `inet`).

![Output del comando ip addr, mostrando la interfaz enp0s1 con dirección
IP asignada](/img/herramientas/macos-arm/ip-addr-ip-asignada.png)

> _¡Mirá, mirá!: tengo internet[^2]_ \- Carlitox

::: danger Ojalá que nunca te pase

Si, después de seguir estos pasos, seguís sin tener conectividad, pasate
[por el foro](https://faq.utnso.com.ar/foro) con bastante paciencia,
porque andá a saber qué anda pasando.

:::

Ahora sí, deberías _más o menos_ tener una VM funcionando. Disfrutala
todo lo que puedas, no te olvides de que esto es todo experimental y que
puede fallar, y tenele mucha paciencia a tus ayudantes con las consultas
que hagas respecto a la VM, porque vamos a tratar de darte una mano,
pero estamos improvisando acá también.

## ¿Y cómo codeo desde la VM Server?

En el caso de VirtualBox, nosotros les proveemos una VM con interfaz gráfica para poder desarrollar. Pero si intentamos utilizar esa misma probablemente te resulte **insoportablemente lento** (por esto de la emulación).

Pero no te preocupes, ¡hay una opción más que podemos usar! Y esto es la magia de trabajar remotamente por SSH.

> En VS Code vamos a descargar está [extensión](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.vscode-remote-extensionpack)
> 
> Y a continuación[^3]:

Reviewer: ```suggestion
En Visual Studio Code vamos a descargar la extensión [Remote Development](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.vscode-remote-extensionpack), y a continuación[^3]:
```

- Abrimos el Command Palette (⇧⌘P)
- Buscamos la opción `Remote-SSH: Connect to Host...`
- Ponemos `utnso@{la ip enp0s1}`
- Si pregunta que sistema estamos usando ponemos Linux, si pregunta que ssh file seleccionamos el que este en nuestra carpeta de usuario
- Nos va a pedir la contraseña es la del sistema (`utnso`)

> - Y ahora ya podemos continuar con:
> 
>   a. Un [simple proyecto C](/primeros-pasos/primer-proyecto-c), abriendo la carpeta con `File > Open Folder...`, o
> 
>   b. El [TP0](/primeros-pasos/tp0), abriendo el workspace que les proveemos con `File > Open Workspace from File...`.

Reviewer: ```suggestion
- Y ahora ya podemos continuar abriendo un [simple proyecto C](/primeros-pasos/primer-proyecto-c).

::: tip

Podemos abrir una consola remota desde Visual Studio Code haciendo click en `Terminal` > `New Terminal`.

:::

```

anduvo todo bien.

```bash
$ ls -lh
total 14061568
-rwxr-xr-x@ 1 user  staff   6.7G Sep 27  2022 Server.vdi
$ qemu-img convert -f vdi -O qcow2 ./Server.vdi Server.qcow2
$ ls -lh
total 27598976
-rw-r--r--@ 1 user  staff   6.5G Jun 28 01:09 Server.qcow2
-rwxr-xr-x@ 1 user  staff   6.7G Sep 27  2022 Server.vdi
$ 
```

## Crear la máquina virtual en UTM

Abrí UTM y dale a `Create a New Virtual Machine`.

Te va a ofrecer Virtualizar o Emular: como estamos haciendo todo esto
para ejecutar la VM que usa otra arquitectura, dale a Emular.

Elegí ejecutar Linux, y cuando te pida que le des una ISO, elegí Browse
y seleccioná _cualquier archivo_: UTM pretende bootear desde esta ISO
para instalar el sistema operativo de tu nueva máquina virtual, pero acá
ya tenemos un disco con un sistema operativo instalado, así que le
estamos mintiendo un poco al wizard simplemente porque no nos deja crear
la VM sin ISO - pero ya lo vamos a solucionar.

En la ventana siguiente, elegí la arquitectura `x86_64` (el nombre
técnico de "AMD64", que, además, probablemente sea el default de esta
opción), dejá el System que tengas por defecto (podría intentar mentirte

> _un montón_ sobre qué son esas opciones, pero aún no tengo mucha idea y
> prefiero ahorrarnos ese inconveniente), dale 1024 MB de RAM (podés poner

Reviewer: ```suggestion
_un montón_ sobre qué son esas opciones, pero aún no tenemos mucha idea y
preferimos ahorrarnos ese inconveniente), dale 1024 MB de RAM (podés poner
```

Seleccioná la segunda solapa IDE Drive de la barra lateral, y esta vez
el Image type debería ser `Disk Image`. Este es el disco virtual en el
que UTM pretendía instalar nuestro sistema operativo - como ya tenemos
otro preparado, dale `Delete Drive` y confirmá la pregunta.

![Captura de la ventana de Preferencias de UTM, con la solapa IDE Drive
seleccionada, mostrando el disco virtual
creado](/img/herramientas/macos-arm/preferencias-ide-disk-image.png)

::: tip Nota

En otras versiones de UTM el boton de delete esta escondido bien abajo sobre el final de la ventana
![Captura de otra version de UTM donde el boton de delete esta más escondido](/img/herramientas/macos-arm/boton-delete-utm.png)

:::

Por último, tocá el botón de `New...` en la barra lateral para agregar
el disco que convertimos antes. En la ventanita flotante, tocá
`Import...` (es decir: no toques Create), y buscá el archivo `qcow2` que
creaste hace un rato con `qemu-img`.

![Captura de la ventana de Preferencias de UTM, con la opción New Drive
seleccionada, mostrando el botón de
Import](/img/herramientas/macos-arm/preferencias-new.png)

Ahora sí, dale `Save`, y ejecutá la VM dándole al botón de Play. ¡Listo!

> Si la VM no enciende todavia tenemos una posibilidad!
> 
> Vamos a editar la configuracion de la VM de nuevo, esta vez vamos a `System` y cambiamos la configuración de System `Standard PC (Q35 + ICH9, 2009) (pc-q35-7.2)` es una opcion muy parecida a la default
> ![Captura de la ventana de Preferencias de UTM, con la opcion de system deseada](/img/herramientas/macos-arm/preferencias-qemu-system.png)
> 
> Cuando elijan la nueva opcion al cartel de warning le dan aceptar, luego buscamos en la barra lateral la opción QEMU,
> y ahí dentro des-seleccionamos la opción `UEFI Boot`. Esto porque al cambiar de sistema se vuelve a poner el check de UEFI.
> 
> Ahora le damos a play nuevamente. ¿Listo?

Reviewer: ```suggestion
::: tip Si la VM no enciende, ¡todavia tenemos una posibilidad!

Vamos a editar la configuracion de la VM de nuevo, esta vez vamos a `System` y cambiamos la configuración de System `Standard PC (Q35 + ICH9, 2009) (pc-q35-7.2)` es una opción muy parecida a la default
![Captura de la ventana de Preferencias de UTM, con la opcion de system deseada](/img/herramientas/macos-arm/preferencias-qemu-system.png)

Cuando elijan la nueva opcion al cartel de warning le dan aceptar.

Luego, buscamos de nuevo la opción QEMU y ahí dentro volvemos
a des-seleccionar la opción `UEFI Boot`. Esto es porque al
cambiar de sistema se vuelve a poner el check de UEFI.

Ahora le damos a play nuevamente. ¿Listo?

:::
```

## Corregir la configuración de red

Probablemente la configuración de red de la VM no te funcione, porque
QEMU seguramente le asigne a la placa de red un nombre de dispositivo
distinto al que tenía el disco que bajaste.

Para corregirlo, iniciá sesión en la VM, ejecutá `ip addr`, y fijate si
la entrada con nombre del estilo `enp0s1` (quizá ese mismo) tiene una
línea que comience con `inet` y una IP. Si está: fiesta, tenés internet,
ganaste.

> 

Reviewer: Agrego la captura antes así se entiende de dónde leer

```suggestion

![Output del comando ip addr. remarcando el nombre de la interfaz de red](/img/herramientas/macos-arm/ip-addr-interfaz.png){data-zoomable}

```

Si la interfaz de red no está configurada, editá el archivo
`/etc/netplan/00-installer-config.yml` con tu editor favorito (¡acordate
de usar `sudo`!), y reemplazá en la línea que dice `enp0s3:` el nombre
de dispositivo que te haya listado `ip addr` (_¡dejale el `:` al final,

> eh!_).
> 
> El archivo yaml tiene un contenido como el siguiente:
> ```yaml
> network:
>     ethernets:
>         enp0s3:
>             dhcp4: true
>     version: 2
> ```
> Tenemos que cambiar el enp0s3 al que nos diga nuestro `ip addr`
> 
> El nombre de la interfaz que nos interesa es la segunda, es posible que en su caso tambien sea `enp0s1`
> ![Output del comando ip addr. remarcando el nombre de la interfaz de red](/img/herramientas/macos-arm/ip-addr-interfaz.png){data-zoomable}
> 
> Finalmente el archivo quedaria de la siguiente forma:

Reviewer: ```suggestion
eh!_):
```

```yaml
network:
    ethernets:

>         enp0s1:

Reviewer: Esto le da formato tipo git diff :3

```suggestion
        enp0s3: // [!code --]
        enp0s1: // [!code ++]
```

porque andá a saber qué anda pasando.

:::

Ahora sí, deberías _más o menos_ tener una VM funcionando. Disfrutala
todo lo que puedas, no te olvides de que esto es todo experimental y que
puede fallar, y tenele mucha paciencia a tus ayudantes con las consultas
que hagas respecto a la VM, porque vamos a tratar de darte una mano,
pero estamos improvisando acá también.

## ¿Y para la Desktop?

Deberías poder seguir _más o menos_ los mismos pasos para instalar la VM
Desktop, pero también probablemente ejecute insoportablemente lento (por
esto de la emulación).

Si probás, contanos qué onda.

## ¿Y si estoy usando Linux/Windows con ARM?

No probamos aún. Deberías tener QEMU disponible en ambos sistemas
operativos, y deberías poder hacer la misma conversión, y deberías poder
armar un comando de `qemu` para ejecutar la VM como hicimos acá - pero
andá a saber cuál es ese comando, y todas las variables que haya en
juego. Como siempre: podés probar, jugar, y pedirnos una mano, pero va a
ser difícil que nos podamos hacer el tiempo (¡y el hardware!) para
probar y entender la situación.

Una vez más, si probás, contanos qué onda.

> ## SSH
> 
> Si la VM grafica ejecuta insoportablemente lento hay una opción más que podemos usar! Y esto es la magia de SSH con la VM Server

Reviewer: ```suggestion
## ¿Y cómo codeo desde la VM Server?

En el caso de VirtualBox, nosotros les proveemos una VM con interfaz gráfica para poder desarrollar. Pero si intentamos utilizar esa misma probablemente te resulte **insoportablemente lento** (por esto de la emulación).

Pero no te preocupes, ¡hay una opción más que podemos usar! Y esto es la magia de trabajar remotamente por SSH.
```

En VS Code vamos a descargar está [extensión](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.vscode-remote-extensionpack)

Y a continuación[^3]:

- Abrimos el Command Palette (⇧⌘P)
- Buscamos la opción `Remote-SSH: Connect to Host...`
- Ponemos `utnso@{la ip enp0s1}`
- Si pregunta que sistema estamos usando ponemos Linux, si pregunta que ssh file seleccionamos el que este en nuestra carpeta de usuario
- Nos va a pedir la contraseña es la del sistema (`utnso`)

> - Y ahora `File > Open Workspace...`
> 
> ::: tip Nota
> 
> Cuando quieras abrir el workspace primero tiene que haber un workspace... si seguis la guia del [tp0](/primeros-pasos/tp0) vas a tener un workspace para usar!
> 
> :::

Reviewer: ```suggestion
- Y ahora:
  a. Si estamos en un [simple proyecto C](/primeros-pasos/primer-proyecto-c), `File > Open File Or Folder...`.
  b. Si estamos en el [TP0](/primeros-pasos/tp0) o el propio repo del TP, abrimos el workspace con `File > Open Workspace...`.
```
